### DISCUSSION TITLE
Logic code in shortcodes or not in shortcodes, that's the question...

### DISCUSSION THREAD

**Discussion by rica-carv** [e107inc/e107#5322] [Q&A Core]
I'm wondering, should logic be in shortcodes or not?
For instance, should the shortcodes only be a "dumb" and flat method to do the output, or it should perform more complicated calculations and have all the functions needed for the shortcode output?

What do you think, or what should be the guidelines for a shortcode coding?

**Comment by Jimmi08** [comment]
Combination.
Each value should have a shortcode for outputting only the value itself. Or with parameter raw...  but you need to know the parameters to get the correct output.
Plus there are shortcodes that are needed for your plugin.

Global shortcodes (theme, e_shortcode) should output elements like news grids, the latest news, or anything else used globally.  

The rule should be - if I need any value, there should be way how to get it without HTML markup, or HTML markup can vary according parameters.  

I use it this way:
- single values (your dumb method) - in shortcodes/batch/ shortcodes
- plugin custom shortcodes (with functionality) - in plugin_shortcodes.php
- plugin elements - in e_shortcode.php

The funny thing is that you don't need shortcode files anymore - look at navigation shortcodes or some Cameron plugins.

**Reply by rica-carv** [reply]
Mainly what i was asking is if the shortcode should be something like this
```
	function sc_db_size(){
		return $this->var['db_size'];
	}
```
or like this (or even more complicated)
```
	function sc_total_views()
	{
		$total_views = 0;
		$sql = e107::getDb();
		if ($sql->gen('SELECT sum(thread_views) AS total FROM `#forum_thread` '))
		{
			$row = $sql->fetch();
			$total_views = $row['total'];
		}
		return $total_views;
	}
```

Even with the first ideia, you also have the option to have that variable generated inside the shortcode, or sent to it via shortcode vars (as it is in the example).
So what i was wondering, is if it could be that the main file should only be used to "stick" together the shortcode and the template, and leave all the logic needed for the shortcodes output inside the shortcode file?
Or leave all the code inside the main file, and leave the shortcode file lightweight?

You will always need the shortcodes files, or you can't override them in your theme, although you can't access the shortcode variables on the custom theme shortcodes (that for me is a major drawback...)

**Reply by Vodhin** [reply]
Let your shortcode do what is needed: some of my shortcode is simple `return $this->var['data']` while other functions get very complicated, even calling one shortcode from another (useful for shortcode overrides )

**Reply by rica-carv** [reply]
@Vodhin Ok, no discussion there.
But what i was trying to see is what's probably best practice and best performance:
- Get all the shortcode dependet logic (ie, variables assignment and calculation code) inside the shortcode file.
- Or leave all that in the main php file, and assign variables to passed and used by the shortcode?
